src/lsp4clj/json_rpc.clj

```diff
@@ -0,0 +1,127 @@
+(ns lsp4clj.json-rpc
+  "Models LSP JSON-RPC as core.async channels of messages (Clojure hashmaps).
+
+  https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#baseProtocol"
+  (:require
+   [camel-snake-kebab.core :as csk]
+   [camel-snake-kebab.extras :as cske]
+   [cheshire.core :as json]
+   [clojure.core.async :as async]
+   [clojure.string :as string]))
+
+(set! *warn-on-reflection* true)
+
+(defn ^:private read-n-bytes [^java.io.InputStream input content-length charset-s]
+  (let [buffer (byte-array content-length)]
+    (loop [total-read 0]
+      (when (< total-read content-length)
+        (let [new-read (.read input buffer total-read (- content-length total-read))]
+          (when (< new-read 0)
+            ;; TODO: return nil instead?
+            (throw (java.io.EOFException.)))
+          (recur (+ total-read new-read)))))
+    (String. ^bytes buffer ^String charset-s)))
+
+(defn ^:private parse-header [line headers]
+  (let [[h v] (string/split line #":\s*" 2)]
+    (assoc headers h v)))
+
+(defn ^:private parse-charset [content-type]
+  (or (when content-type
+        (when-let [[_ charset] (re-find #"(?i)charset=(.*)$" content-type)]
+          (when (not= "utf8" charset)
+            charset)))
+      "utf-8"))
+
+(defn ^:private read-message [input headers keyword-function]
+  (try
+    (let [content-length (parse-long (get headers "Content-Length"))
+          charset-s (parse-charset (get headers "Content-Type"))
+          content (read-n-bytes input content-length charset-s)]
+      (json/parse-string content keyword-function))
+    (catch Exception _
+      :parse-error)))
+
+(defn ^:private kw->camelCaseString
+  "Convert keywords to camelCase strings, but preserve capitalization of things
+  that are already strings."
+  [k]
+  (cond-> k (keyword? k) csk/->camelCaseString))
+
+(def ^:private write-lock (Object.))
+
+(defn ^:private write-message [^java.io.OutputStream output msg]
+  (let [content (json/generate-string (cske/transform-keys kw->camelCaseString msg))
+        content-bytes (.getBytes content "utf-8")]
+    (locking write-lock
```

Comment:
There was some issues in the past on Clojure with the locking macro on Graalvm images, I think it was solved, but it's good to keep in mind that case during tests with graalvm, I do think integration tests cover that though

Reply:
I read a little about that bug—yikes. Seems nasty. It sounds like it's fixed though. The GraalVM build works locally for me on a Mac, and as you said, if it were broken for Linux we'd see that in the integration tests.